### PR TITLE
feat: cheatcode to read and set environment variables

### PR DESCRIPTION
**Pull request by 0xYYY**
## Motivation

Closes https://github.com/foundry-rs/foundry/issues/1691.

## Solution

- Add `env{Bool,Uint,Int,Address,Bytes32,String,Bytes}` cheatcodes to read environment variables.

  Each of these cheatcodes has two variants, `envTYPE(string env_key)` and `envTYPE(string env_key, string delimiter)` where the latter is for reading as arrays. 
  
  For example, with environment variable `strings=hello, world!|thisisastring`, one can use
  ```
  string memory key = "strings";
  string memory delim = "|";
  string[] memory strings = envString(key, delim);
  ```
  to access it. And `strings[0]` will have value `hello, world!` and `strings[1]` will have value `thisisastring`. 

- Add `setEnv` cheatcode to set environment variables, mainly for writing tests for above cheatcodes.

## TODO
- [x]  Add tests.

### CODE REVIEW

File: evm/src/executor/abi.rs [hunk 1]
```diff
@@ -21,6 +21,10 @@ ethers::contract::abigen!(
             store(address,bytes32,bytes32)
             load(address,bytes32)(bytes32)
             ffi(string[])(bytes)
+            env_string(string)(string)
+            env_uint(string)(uint256)
+            env_int(string)(int256)
+            env_bool(string)(bool)
```
Comment (onbjerg) [inline evm/src/executor/abi.rs:27]: These should be camel case as all the other cheatcodes

```suggestion
            env_string(string)(string)
            env_uint(string)(uint256)
            env_int(string)(int256)
            env_bool(string)(bool)
```

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 2]
```diff
@@ -86,6 +121,10 @@ pub fn apply(ffi_enabled: bool, call: &HEVMCalls) -> Option<Result<Bytes, Bytes>
             }
         }
         HEVMCalls::GetCode(inner) => get_code(&inner.0),
+        HEVMCalls::EnvBool(inner) => env_bool(&inner.0),
+        HEVMCalls::EnvUint(inner) => env_uint(&inner.0),
+        HEVMCalls::EnvInt(inner) => env_int(&inner.0),
+        HEVMCalls::EnvString(inner) => env_string(&inner.0),
```
Comment (mds1) [inline evm/src/executor/inspector/cheatcodes/ext.rs:127]: Could we also add `address`, bytes`, and `bytes32` types?

**Comment by mattsse** [comment]
just an idea but would overloading them with an additional string param makes sense that then returns an array? such as `env_int("VAR", ",")(int[])`

File: evm/src/executor/abi.rs (hunk 1 above)
Comment (0xYYY) [reply evm/src/executor/abi.rs]: thanks for catching this, fixed

File: evm/src/executor/inspector/cheatcodes/ext.rs (hunk 2 above)
Comment (0xYYY) [reply evm/src/executor/inspector/cheatcodes/ext.rs:146]: great suggestion, i've added them

**Comment by 0xYYY** [comment]
> just an idea but would overloading them with an additional string param makes sense that then returns an array? such as `env_int("VAR", ",")(int[])`

another idea could be that we always return an array (for types other than `string` and `bytes`), between these 2 options, i think it come down to whether
```
int a = envInt("VAR");
int b[] = envInt("VAR", ",");
```
or
```
int a = envInt("VAR")[0];
int b[] = envInt("VAR");
```
is a better pattern/UX for devs.

**Comment by 0xYYY** [comment]
> just an idea but would overloading them with an additional string param makes sense that then returns an array? such as `env_int("VAR", ",")(int[])`

perhaps just having another function with different name is more straightforward, e.g. `env_int_arr(string)(int[])`, since users will have to use some kind of specifier anyway, having that in the function name instead of in parameters may be clearer

with the overloading approach, there may be some confusion like "what happers if i provide other string in the 2nd param"

wdyt?

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 3]
```diff
@@ -76,6 +78,116 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
 }
 
+// struct BytesResult(Result<Bytes, Bytes>);
+
+// impl<T: AbiEncode, E: ToString> From<Result<T, E>> for BytesResult {
+//     fn from(src: Result<T, E>) -> Self {
+//         BytesResult(src.map(|v| v.encode().into()).map_err(|e| e.to_string().encode().into()))
+//     }
+// }
+
```
Comment (gakonst) [inline evm/src/executor/inspector/cheatcodes/ext.rs:88]: remove
```suggestion
```

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 4]
```diff
@@ -76,6 +78,116 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+fn set_env(key: &str, val: &str) -> Result<Bytes, Bytes> {
+    // `std::env::set_var` may panic in the following situations
+    // ref: https://doc.rust-lang.org/std/env/fn.set_var.html
+    if key.is_empty() {
+        Err("Environment variable key can't be empty".to_string().encode().into())
+    } else if key.contains('=') {
+        Err("Environment variable key can't contain equal sign `=`".to_string().encode().into())
+    } else if key.contains('\0') {
+        Err("Environment variable key can't contain NUL character `\0`".to_string().encode().into())
```
Comment (gakonst) [inline evm/src/executor/inspector/cheatcodes/ext.rs:97]:
```suggestion
        Err("Environment variable key can't contain NULL character `\0`".to_string().encode().into())
```

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 5]
```diff
@@ -76,6 +78,116 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+    } else if val.contains('\0') {
+        Err("Environment variable value can't contain NUL character `\0`"
```
Comment (gakonst) [inline evm/src/executor/inspector/cheatcodes/ext.rs:99]:
```suggestion
        Err("Environment variable value can't contain NULL character `\0`"
```

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 6]
```diff
@@ -76,6 +78,116 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+            .encode()
+            .into())
+    } else {
+        env::set_var(key, val);
+        Ok(Bytes::new())
+    }
+}
+
+fn get_env(key: &str, _type: &str, is_array: bool) -> Result<Bytes, Bytes> {
+    // TODO: does it make sense to have `_type == "bytes" && is_array == true`?
```
Comment (gakonst) [inline evm/src/executor/inspector/cheatcodes/ext.rs:110]: wdym? the way it's done now seems fine i think. let's make `_type` -> `r#type`?

File: evm/src/executor/inspector/cheatcodes/ext.rs (hunk 4 above)
Comment (0xYYY) [reply evm/src/executor/inspector/cheatcodes/ext.rs]: Was following `std::env::set_var` doc which uses the term *NUL character*. [ref](https://doc.rust-lang.org/std/env/fn.set_var.html)

In Rust reference on Tokens, it also states that "The *null escape* is the character `U+0030` (`0`) and denotes the Unicode value `U+0000` (*NUL*)". [ref](https://doc.rust-lang.org/reference/tokens.html)

File: evm/src/executor/inspector/cheatcodes/ext.rs (hunk 6 above)
Comment (0xYYY) [reply evm/src/executor/inspector/cheatcodes/ext.rs]: fixed

File: evm/src/executor/inspector/cheatcodes/ext.rs (hunk 3 above)
Comment (0xYYY) [reply evm/src/executor/inspector/cheatcodes/ext.rs]: removed

**Comment by gakonst** [comment]
@0xYYY let us know when this is good for review!

**Comment by 0xYYY** [comment]
@gakonst ready for review, please take a look, thanks
cc @mattsse @onbjerg @mds1

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 7]
```diff
@@ -76,6 +78,80 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
     }
 }
 
+fn set_env(key: &str, val: &str) -> Result<Bytes, Bytes> {
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:81]: til that set_env panics

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 8]
```diff
@@ -76,6 +78,80 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+            .encode()
+            .into())
+    } else if val.contains('\0') {
+        Err("Environment variable value can't contain NUL character `\\0`"
+            .to_string()
+            .encode()
+            .into())
+    } else {
+        env::set_var(key, val);
+        Ok(Bytes::new())
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:100]: I don't think we have a custom `bail!` macro yet that does 
`return Err($msg.to_string().encode().into())`

would be a nice to have imo

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 9]
```diff
@@ -76,6 +78,80 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+        if v.starts_with("0x") {
+            let v = Vec::from_hex(v.strip_prefix("0x").unwrap()).map_err(|e| e.to_string())?;
+            Ok(U256::from_little_endian(&v))
+        } else {
+            U256::from_dec_str(v).map_err(|e| e.to_string())
+        }
+    };
+    let parse_int = |v: &str| {
+        // hex string may start with "0x", "+0x", or "-0x"
+        if v.contains("0x") {
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:123]: should only check the start, not `contains`, right?

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 10]
```diff
@@ -76,6 +78,80 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+            "address" => parse_address(v).map(Token::Address).map_err(|e| e.to_string()),
+            "bytes32" => parse_bytes(v).map(Token::FixedBytes).map_err(|e| e.to_string()),
+            "string" => parse_string(v).map(Token::String).map_err(|_| "".to_string()),
+            "bytes" => parse_bytes(v).map(Token::Bytes).map_err(|e| e.to_string()),
+            _ => Err(format!("{} is not a supported type", r#type)),
+        })
+        .collect::<Result<Vec<Token>, String>>()
+        .map(|tokens| {
+            if delim.is_none() {
+                abi::encode(&[tokens[0].clone()]).into()
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:147]: can use `remove(0)` here

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 11]
```diff
@@ -76,6 +78,80 @@ fn get_code(path: &str) -> Result<Bytes, Bytes> {
...
+            .to_string()
+            .encode()
+            .into())
+    } else {
+        env::set_var(key, val);
+        Ok(Bytes::new())
+    }
+}
+
+fn get_env(key: &str, r#type: &str, delim: Option<&str>) -> Result<Bytes, Bytes> {
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:104]: instead of type as `str` we could use `ParamType` here

**Review by mattsse — APPROVED** [review]
lgtm, smol nits

**Review by onbjerg — APPROVED** [review]
LGTM (see Matt nits)

File: evm/src/executor/inspector/cheatcodes/ext.rs (hunk 8 above)
Comment (0xYYY) [reply evm/src/executor/inspector/cheatcodes/ext.rs:100]: would probably do this in a separate PR, to deal with this boilerplate code scattered around the `cheatcodes/` codebase

**Pull request merged**